### DISCUSSION TITLE
The best practices for using a cache instance

### DISCUSSION THREAD

**Discussion by vainivei** [cache2k/cache2k#191] [Q&A] [answered]
This might be a really dumb question, but what are the best practices for using a cache instance. In all the examples I've found online, the Cache object is created on class creation as a regular variable, meaning that each instance creates a new Cache instance (or throws 'IllegalArgumentException: Cache already created' if the cache is named). So the classic options are that the cache instance is either stored in a final static variable or wrapped inside a Singleton implementation. Are these best practices or do you have better options?

**Comment by cruftex** [comment, marked as answer]
In the example you see we assume a singleton pattern and a service class typically implementing some kind of DAO methods.

It depends on the environment or frameworks you use. Modern applications typically use some kind of IoC container (e.g. Spring or Guice) which takes care that the resources are initialized once.

Note that its also possible to the configuration in an external file. If you don't use the programmatic configuration and creation you are able to just retrieve a cache from the cache manager. Then the cache manager takes care that it is initialized once. This happens under the hood if you configure cache2k to be used with hibernate, for example.

If you tell me more about your application and environment I can add more details or an example, if needed.
